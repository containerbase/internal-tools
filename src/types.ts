export enum Commands {
  BinaryBuilder = 'binary-builder',
  DockerConfig = 'docker-config',
  DockerBuilder = 'docker-builder',
}
