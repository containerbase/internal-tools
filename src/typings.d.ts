/* eslint-disable  */
declare module 'www-authenticate' {
  namespace wwwAuthenticate {
    namespace parsers {
      class WWW_Authenticate {
        readonly parms: Record<string, string>;
        constructor(auth: string);
      }
    }
  }
  export = wwwAuthenticate;
}

declare namespace NodeJS {
  interface Global {
    renovateCache: unknown;
    repoCache: unknown;
  }
}
