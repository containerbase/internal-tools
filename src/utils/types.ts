export type ExecResult = {
  readonly code: number;
  readonly stdout: string;
  readonly stderr: string;
};

export class ExecError extends Error implements ExecResult {
  readonly name = 'ExecError';

  constructor(
    public readonly code: number,
    public readonly stdout: string,
    public readonly stderr: string,
    public readonly cmd: string
  ) {
    super(`ExecError: (${code}) ` + stderr.split('\n').slice(-10).join('\n'));
  }
}

export type ConfigFile = {
  datasource: string;
  image: string;
  depName?: string;
  versioning?: string;
  startVersion: string;
  cache?: string;
  buildArg?: string;
  ignoredVersions?: string[];
  forceUnstable?: boolean;
  versions?: string[];
  latestVersion?: string;
  maxVersions?: number;
  extractVersion?: string;
};

export type Config = {
  buildArg: string;
  buildArgs: string[];
  buildOnly: boolean;
  tagSuffix?: string;
  depName: string;
  imagePrefix: string;
  image: string;
  ignoredVersions: string[];
  majorMinor: boolean;
  lastOnly: boolean;
  dryRun: boolean;
  prune: boolean;
  platforms?: string[];
} & ConfigFile;

export type BinaryBuilderConfig = {
  buildArgs: string[];
  depName: string;
  ignoredVersions: string[];
  lastOnly: boolean;
  dryRun: boolean;
} & ConfigFile;
