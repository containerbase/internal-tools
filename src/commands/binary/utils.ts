import { getInput } from '@actions/core';
import is from '@sindresorhus/is';
import { getDefaultVersioning } from 'renovate/dist/modules/datasource';
import { getArch, getArg, getDistro, isDryRun, readJson } from '../../util';
import { readDockerConfig } from '../../utils/config';
import {
  DockerPlatform,
  dockerBuildx,
  dockerRun,
} from '../../utils/docker/common';
import log from '../../utils/logger';
import type { BinaryBuilderConfig, ConfigFile } from '../../utils/types';

export async function getConfig(): Promise<BinaryBuilderConfig> {
  const configFile = getInput('config') || 'builder.json';

  const cfg = await readJson<ConfigFile>(configFile);

  if (!is.object(cfg)) {
    throw new Error('missing-config');
  }

  if (!is.string(cfg.image)) {
    cfg.image = getInput('image', { required: true });
  }

  if (!is.string(cfg.buildArg)) {
    cfg.buildArg = cfg.image.toUpperCase() + '_VERSION';
  }

  await readDockerConfig(cfg);

  return {
    ...cfg,
    ignoredVersions: cfg.ignoredVersions ?? [],
    dryRun: isDryRun(),
    lastOnly: getArg('last-only') == 'true',
    buildArgs: getArg('build-args', { multi: true }),
    versioning: cfg.versioning ?? getDefaultVersioning(cfg.datasource),
  } as BinaryBuilderConfig;
}

export async function createBuilderImage(
  ws: string,
  { buildArgs }: BinaryBuilderConfig
): Promise<void> {
  log('Creating builder image');
  const args = [
    'build',
    '--load',
    '-t',
    'builder',
    '--build-arg',
    `DISTRO=${getDistro()}`,
  ];
  const arch = getArch();
  if (is.nonEmptyString(arch)) {
    args.push('--platform', DockerPlatform[arch]);
  }
  if (is.nonEmptyArray(buildArgs)) {
    args.push(...buildArgs.map((b) => `--build-arg=${b}`));
  }
  await dockerBuildx(...args, ws);
}

export async function runBuilder(ws: string, version: string): Promise<void> {
  const args = ['--name', 'builder', '--volume', `${ws}/.cache:/cache`];
  const arch = getArch();
  if (is.nonEmptyString(arch)) {
    args.push('--platform', DockerPlatform[arch]);
  }
  await dockerRun(...args, 'builder', version);
}
