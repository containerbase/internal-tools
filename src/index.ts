// Node v16 is missing structuredClone, so we polyfill it here.
import 'core-js/actual/structured-clone';
import run from './runner';

run().catch(console.error);
