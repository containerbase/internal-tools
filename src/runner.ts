import { getInput, setFailed } from '@actions/core';
import * as chalk from 'chalk';
import { Commands } from './types';
import log from './utils/logger';

export default async function run(): Promise<void> {
  try {
    log.info(chalk.blue('Renovate Docker Builder'));
    const cmd = getInput('command') as Commands;
    log.info(chalk.yellow('Executing:'), ` ${cmd}`);
    switch (cmd) {
      case Commands.BinaryBuilder:
        await (await import('./commands/binary')).run();
        break;
      case Commands.DockerBuilder:
        await (await import('./commands/docker/builder')).run();
        break;

      case Commands.DockerConfig:
        await (await import('./commands/docker/config')).run();
        break;

      default:
        log.error(chalk.red('Unknown command:'), cmd);
        break;
    }
  } catch (error) {
    setFailed((error as Error).message);
  }
}
