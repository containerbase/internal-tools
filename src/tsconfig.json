{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "rootDir": ".",
    "outDir": "dist",
    "types": ["node"]
  }
}
