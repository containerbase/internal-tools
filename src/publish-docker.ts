import { getInput } from '@actions/core';
import { exec } from './util';
import chalk from 'chalk';
import log from 'fancy-log';
import { getRemoteImageId, getLocalImageId } from './utils/docker';

export const MultiArgsSplitRe = /\s*(?:;|$)\s*/;
export async function run(): Promise<void> {
  const dryRun = !!getInput('dry-run');
  const image = getInput('image');
  let tags = getInput('tags')
    ?.split(MultiArgsSplitRe)
    .filter(Boolean);

  if (!tags?.length) {
    tags = ['latest'];
  }

  if (!image) {
    throw new Error('Missing image');
  }

  for (const tag of tags) {
    const fullName = `${image}:${tag}`;
    log.info(chalk.blue('Processing image:'), chalk.yellow(fullName));

    log('Fetch new id');
    const newId = await getLocalImageId(image, tag);

    log('Fetch old id');
    const oldId = await getRemoteImageId(image, tag);

    if (oldId === newId) {
      log('Image uptodate, no push nessessary: ', chalk.yellow(oldId));
      continue;
    }

    log('Publish new image', `${oldId} > ${newId}`);
    if (dryRun) {
      log.warn(chalk.yellow('DRY-RUN: Would push: '), fullName);
    } else {
      await exec('docker', ['push', fullName]);
    }
    log.info(chalk.blue('Processing image finished: ', newId));
  }
}
