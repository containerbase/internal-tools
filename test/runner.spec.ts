import * as _core from '@actions/core';
import run from '../src/runner';
import { Commands } from '../src/types';
import { mocked } from './utils';

jest.mock('../src/commands/docker/config');
jest.mock('../src/commands/docker/builder');
jest.mock('../src/commands/binary');

const core = mocked(_core);

describe('runner', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('works', async () => {
    core.getInput.mockReturnValueOnce('works');
    await run();

    expect(core.getInput.mock.calls).toMatchSnapshot();
    expect(core.setFailed).not.toHaveBeenCalled();
  });

  it(Commands.BinaryBuilder, async () => {
    core.getInput.mockReturnValueOnce(Commands.BinaryBuilder);
    await run();
    expect(core.getInput).toHaveBeenCalledWith('command');
    expect(core.setFailed).not.toHaveBeenCalled();
  });

  it(Commands.DockerBuilder, async () => {
    core.getInput.mockReturnValueOnce(Commands.DockerBuilder);
    await run();
    expect(core.getInput).toHaveBeenCalledWith('command');
    expect(core.setFailed).not.toHaveBeenCalled();
  });

  it(Commands.DockerConfig, async () => {
    core.getInput.mockReturnValueOnce(Commands.DockerConfig);
    await run();
    expect(core.getInput).toHaveBeenCalledWith('command');
    expect(core.setFailed).not.toHaveBeenCalled();
  });

  it('catch error', async () => {
    core.getInput.mockImplementationOnce((_) => {
      throw new Error('test');
    });
    await run();

    expect(core.getInput.mock.calls).toMatchSnapshot();
    expect(core.setFailed).toHaveBeenCalledWith('test');
  });
});
