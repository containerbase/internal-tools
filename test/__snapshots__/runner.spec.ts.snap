// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runner catch error 1`] = `
Array [
  Array [
    "command",
  ],
]
`;

exports[`runner works 1`] = `
Array [
  Array [
    "command",
  ],
]
`;
